streams.go

```diff
@@ -0,0 +1,296 @@
+	Info ByteStreamInfo
+}
+
+func newByteStreamWriter(info ByteStreamInfo, header *protocol.DataStream_Header, e *RTCEngine, destinationIdentities []string, onProgress *func(progress float64)) *ByteStreamWriter {
+	e.publishStreamHeader(header, destinationIdentities)
+
+	onWrite := func(data []byte) {
+		writeStreamBytes(data, e, info.Id, destinationIdentities, info.Size, onProgress)
+	}
+
+	onClose := func() {
+		e.publishStreamTrailer(info.Id, destinationIdentities)
+	}
+
+	return &ByteStreamWriter{
+		baseStreamWriter: newBaseStreamWriter[[]byte](onWrite, onClose),
+		Info:             info,
+	}
+}
+
+type baseStreamReader struct {
+	reader        chan *protocol.DataStream_Chunk
+	totalByteSize *uint64
+	bytesReceived int
+
+	closed     bool
+	readerLock sync.Mutex
+
+	onProgress *func(progress float64)
+}
+
+func newBaseStreamReader(totalByteSize *uint64) *baseStreamReader {
+	baseReader := &baseStreamReader{
+		reader:        make(chan *protocol.DataStream_Chunk),
+		bytesReceived: 0,
+	}
+	if totalByteSize != nil {
+		baseReader.totalByteSize = totalByteSize
+	}
+	return baseReader
+}
+
+func (r *baseStreamReader) enqueue(chunk *protocol.DataStream_Chunk) {
+	r.readerLock.Lock()
+	defer r.readerLock.Unlock()
+
+	if r.closed {
+		return
+	}
+	r.reader <- chunk
+}
+
+func (r *baseStreamReader) OnProgress(onProgress *func(progress float64)) {
+	r.onProgress = onProgress
+}
+
+func (r *baseStreamReader) close() {
+	r.readerLock.Lock()
+	defer r.readerLock.Unlock()
+
+	if !r.closed {
+		close(r.reader)
+		r.closed = true
+	}
+}
+
+type TextStreamReader struct {
+	*baseStreamReader
+	Info TextStreamInfo
+
+	// Q: what are we using this for?
+	receivedChunks map[uint64]*protocol.DataStream_Chunk
+}
+
+func NewTextStreamReader(info TextStreamInfo, totalChunkCount *uint64) *TextStreamReader {
+	return &TextStreamReader{
+		baseStreamReader: newBaseStreamReader(totalChunkCount),
+		Info:             info,
+		receivedChunks:   make(map[uint64]*protocol.DataStream_Chunk),
+	}
+}
+
+func (r *TextStreamReader) handleChunkReceived(chunk *protocol.DataStream_Chunk) {
+	index := chunk.ChunkIndex
+	previousChunkAtIndex, ok := r.receivedChunks[index]
+
+	if ok && previousChunkAtIndex != nil && previousChunkAtIndex.Version > chunk.Version {
+		// we have a newer version already,dropping the old one
+		return
+	}
+
+	r.receivedChunks[index] = chunk
+	r.bytesReceived += len(chunk.Content)
+
+	if r.totalByteSize != nil && r.onProgress != nil {
+		currentProgress := float64(r.bytesReceived) / float64(*r.totalByteSize)
+		(*r.onProgress)(currentProgress)
+	}
+}
+
+func (r *TextStreamReader) Read() <-chan string {
+	readOnlyChan := make(chan string)
+
+	go func() {
+		defer close(readOnlyChan)
+
+		for chunk := range r.reader {
+			r.handleChunkReceived(chunk)
+			readOnlyChan <- string(chunk.Content)
+		}
+	}()
+
+	return readOnlyChan
+}
+
+func (r *TextStreamReader) ReadAll() string {
+	// Q: what if part of the data is already read through Read()?
```

Comment:
readAll should lock the stream reader. it's either iterate via read or readAll, mixing both is not supported in other impemenations

Reply:
Got it, thanks @lukasIO!

Reply:
In the current implementation with Buffer, the user can use a mix of read methods, I think it's better that way. ReadAll() here returns what is unread.

@boks1971 what do you think?

Reply:
Not sure about mixed methods. Need to think about it more.